lib/spack/spack/database.py

```diff
@@ -1144,10 +1141,9 @@ def _add(
             new_spec._full_hash = spec._full_hash
 
         else:
-            # If it is already there, mark it as installed and update
-            # installation time
-            self._data[key].installed = True
-            self._data[key].installation_time = _now()
+            # It is already in the database
+            self._data[key].installed = installed
+            self._data[key].installation_time = installation_time
```

Comment:
This needs to only update the installation time if the `installed` value changed.

```suggestion
            new_install = self._data[key].installed != installed
            self._data[key].installed = installed
            if new_install:
                self._data[key].installation_time = installation_time
```

Reply:
I think this codepath is hit for overwrite installs (like develop specs), in which case you'd want to "touch" the record to reflect the overwrite time, right?

If it goes from installed -> not installed, then probably you don't want the installation_time to be modified.

Reply:
I'm pretty sure this code path is also hit for

```
spack install bar
spack install foo^bar  # hits this code path for bar
```

But you're right, we need something cleverer than my suggestion for the overwrite case.

Reply:
Hm okay, right now the behavior is effectively unchanged if the install time is not passed as an argument. Maybe getting install times right could work as a follow up? Seems like a few code paths have to be considered... (install-as-dependency, overwrite, re-registering through reindex...)

Reply:
We could just stat the `spec.prefix` directory for its creation time, and only update the installation time for previously installed packages if that time is greater than the existing installation time.

```suggestion
            set_installation_time = installed and not self._data[key].installed
            if self._data[key].installed and installed:
                # On UNIX, creation times are really metadata modification times.
                # shouldn't make a difference in this case
                prefix_creation_time = os.stat(spec.prefix).st_ctime
                if prefix_creation_time > self._data[key].installation_time:
                    set_installation_time = True
                   
            self._data[key].installed = installed
            if set_installation_time:
                self._data[key].installation_time = installation_time
```

Reply:
But I agree that this PR doesn't make this particular issue worse, so if you decide it needs to be pushed to a next PR I'm fine with that.

Reply:
I'm a bit reluctant to add this, considering `installation_time` is relied on in `installer.py`:

```
            if installed_in_db and (
                    dep.dag_hash() not in request.overwrite or
                    rec.installation_time > request.overwrite_time):
```

and I can already foresee hard to debug issues with distributed builds etc. What I'd like is to have clear bug fixes and backport it.